File: concepts/05-trees/node/redBlackTree.js
/**
 * Red-Black Tree Implementation in Node.js
 */

class RBNode {
    /**
     * Node for Red-Black Tree
     */
    static RED = true;
    static BLACK = false;

    constructor(val) {
        this.val = val;
        this.color = RBNode.RED;
        this.left = null;
        this.right = null;
        this.parent = null;
    }
}

class RedBlackTree {
    /**
     * Red-Black Tree - Self-balancing BST
     */
    constructor() {
        this.nil = new RBNode(0);
        this.nil.color = RBNode.BLACK;
        this.root = this.nil;
    }

    /**
     * Left rotation
     */
    leftRotate(x) {
        const y = x.right;
        x.right = y.left;

        if (y.left !== this.nil) {
            y.left.parent = x;
        }

        y.parent = x.parent;

        if (x.parent === this.nil) {
            this.root = y;
        } else if (x === x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }

        y.left = x;
        x.parent = y;
    }

    /**
     * Right rotation
     */
    rightRotate(x) {
        const y = x.left;
        x.left = y.right;

        if (y.right !== this.nil) {
            y.right.parent = x;
        }

        y.parent = x.parent;

        if (x.parent === this.nil) {
            this.root = y;
        } else if (x === x.parent.right) {
            x.parent.right = y;
        } else {
            x.parent.left = y;
        }

        y.right = x;
        x.parent = y;
    }

    /**
     * Insert value with RB tree properties
     */
    insert(val) {
        const z = new RBNode(val);
        z.left = this.nil;
        z.right = this.nil;

        let y = this.nil;
        let x = this.root;

        while (x !== this.nil) {
            y = x;
            if (z.val < x.val) {
                x = x.left;
            } else {
                x = x.right;
            }
        }

        z.parent = y;

        if (y === this.nil) {
            this.root = z;
        } else if (z.val < y.val) {
            y.left = z;
        } else {
            y.right = z;
        }

        z.color = RBNode.RED;
        this._insertFixup(z);
    }

    /**
     * Fix RB tree properties after insertion
     */
    _insertFixup(z) {
        while (z.parent.color === RBNode.RED) {
            if (z.parent === z.parent.parent.left) {
                const y = z.parent.parent.right;
                if (y.color === RBNode.RED) {
                    z.parent.color = RBNode.BLACK;
                    y.color = RBNode.BLACK;
                    z.parent.parent.color = RBNode.RED;
                    z = z.parent.parent;
                } else {
                    if (z === z.parent.right) {
                        z = z.parent;
                        this.leftRotate(z);
                    }
                    z.parent.color = RBNode.BLACK;
                    z.parent.parent.color = RBNode.RED;
                    this.rightRotate(z.parent.parent);
                }
            } else {
                const y = z.parent.parent.left;
                if (y.color === RBNode.RED) {
                    z.parent.color = RBNode.BLACK;
                    y.color = RBNode.BLACK;
                    z.parent.parent.color = RBNode.RED;
                    z = z.parent.parent;
                } else {
                    if (z === z.parent.left) {
                        z = z.parent;
                        this.rightRotate(z);
                    }
                    z.parent.color = RBNode.BLACK;
                    z.parent.parent.color = RBNode.RED;
                    this.leftRotate(z.parent.parent);
                }
            }
        }

        this.root.color = RBNode.BLACK;
    }

    /**
     * Search for value
     */
    search(val) {
        let x = this.root;
        while (x !== this.nil) {
            if (val === x.val) {
                return x;
            } else if (val < x.val) {
                x = x.left;
            } else {
                x = x.right;
            }
        }
        return null;
    }

    /**
     * Inorder traversal
     */
    inorder() {
        const result = [];

        const traverse = (node) => {
            if (node !== this.nil) {
                traverse(node.left);
                result.push([node.val, node.color === RBNode.RED ? "RED" : "BLACK"]);
                traverse(node.right);
            }
        };

        traverse(this.root);
        return result;
    }

    /**
     * Get black height of tree
     */
    getBlackHeight() {
        const countBlack = (node) => {
            if (node === this.nil) {
                return 0;
            }
            const leftBlack = countBlack(node.left);
            if (leftBlack === -1) {
                return -1;
            }
            const rightBlack = countBlack(node.right);
            if (rightBlack === -1) {
                return -1;
            }

            if (leftBlack !== rightBlack) {
                return -1;
            }

            return leftBlack + (node.color === RBNode.RED ? 0 : 1);
        };

        return countBlack(this.root);
    }
}

// Example usage
if (require.main === module) {
    const rbt = new RedBlackTree();

    const values = [7, 3, 18, 10, 22, 8, 11, 26];
    values.forEach(val => rbt.insert(val));

    console.log("Red-Black Tree Inorder Traversal (value, color):");
    console.log(rbt.inorder());

    console.log("\nSearch 10:", rbt.search(10) !== null);
    console.log("Search 99:", rbt.search(99) !== null);

    console.log("\nBlack Height:", rbt.getBlackHeight());
}

module.exports = { RedBlackTree };

File: concepts/05-trees/python/red_black_tree.py
"""
Red-Black Tree Implementation in Python
"""

class RBNode:
    """Node for Red-Black Tree"""
    RED = True
    BLACK = False
    
    def __init__(self, val):
        self.val = val
        self.color = RBNode.RED
        self.left = None
        self.right = None
        self.parent = None


class RedBlackTree:
    """Red-Black Tree - Self-balancing BST"""
    
    def __init__(self):
        self.nil = RBNode(0)
        self.nil.color = RBNode.BLACK
        self.root = self.nil
    
    def left_rotate(self, x):
        """Left rotation"""
        y = x.right
        x.right = y.left
        
        if y.left != self.nil:
            y.left.parent = x
        
        y.parent = x.parent
        
        if x.parent == self.nil:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y
    
    def right_rotate(self, x):
        """Right rotation"""
        y = x.left
        x.left = y.right
        
        if y.right != self.nil:
            y.right.parent = x
        
        y.parent = x.parent
        
        if x.parent == self.nil:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        
        y.right = x
        x.parent = y
    
    def insert(self, val):
        """Insert value with RB tree properties"""
        z = RBNode(val)
        z.left = self.nil
        z.right = self.nil
        
        y = self.nil
        x = self.root
        
        while x != self.nil:
            y = x
            if z.val < x.val:
                x = x.left
            else:
                x = x.right
        
        z.parent = y
        
        if y == self.nil:
            self.root = z
        elif z.val < y.val:
            y.left = z
        else:
            y.right = z
        
        z.color = RBNode.RED
        self._insert_fixup(z)
    
    def _insert_fixup(self, z):
        """Fix RB tree properties after insertion"""
        while z.parent.color == RBNode.RED:
            if z.parent == z.parent.parent.left:
                y = z.parent.parent.right
                if y.color == RBNode.RED:
                    z.parent.color = RBNode.BLACK
                    y.color = RBNode.BLACK
                    z.parent.parent.color = RBNode.RED
                    z = z.parent.parent
                else:
                    if z == z.parent.right:
                        z = z.parent
                        self.left_rotate(z)
                    z.parent.color = RBNode.BLACK
                    z.parent.parent.color = RBNode.RED
                    self.right_rotate(z.parent.parent)
            else:
                y = z.parent.parent.left
                if y.color == RBNode.RED:
                    z.parent.color = RBNode.BLACK
                    y.color = RBNode.BLACK
                    z.parent.parent.color = RBNode.RED
                    z = z.parent.parent
                else:
                    if z == z.parent.left:
                        z = z.parent
                        self.right_rotate(z)
                    z.parent.color = RBNode.BLACK
                    z.parent.parent.color = RBNode.RED
                    self.left_rotate(z.parent.parent)
        
        self.root.color = RBNode.BLACK
    
    def search(self, val):
        """Search for value"""
        x = self.root
        while x != self.nil:
            if val == x.val:
                return x
            elif val < x.val:
                x = x.left
            else:
                x = x.right
        return None
    
    def inorder(self):
        """Inorder traversal"""
        result = []
        
        def traverse(node):
            if node != self.nil:
                traverse(node.left)
                result.append((node.val, "RED" if node.color == RBNode.RED else "BLACK"))
                traverse(node.right)
        
        traverse(self.root)
        return result
    
    def get_black_height(self):
        """Get black height of tree"""
        def count_black(node):
            if node == self.nil:
                return 0
            left_black = count_black(node.left)
            if left_black == -1:
                return -1
            right_black = count_black(node.right)
            if right_black == -1:
                return -1
            
            if left_black != right_black:
                return -1
            
            return left_black + (0 if node.color == RBNode.RED else 1)
        
        return count_black(self.root)


# Example usage
if __name__ == "__main__":
    rbt = RedBlackTree()
    
    values = [7, 3, 18, 10, 22, 8, 11, 26]
    for val in values:
        rbt.insert(val)
    
    print("Red-Black Tree Inorder Traversal (value, color):")
    print(rbt.inorder())
    
    print("\nSearch 10:", rbt.search(10) is not None)
    print("Search 99:", rbt.search(99) is not None)
    
    print("\nBlack Height:", rbt.get_black_height())

